### ISSUE TITLE
have scrapped phone number?? from just dial

### ISSUE THREAD

**Issue by dvijparekh** [Alankar0416/Justdial-Scrapper#1]

**Comment by Alankar0416** [comment]
I was able to earlier, but it seems they have started sending svg image instead of numbers.

**Comment by Alankar0416** [comment]
@dvijparekh1995 However we can take the class name and map it from there. But this will break when they change it again.

**Comment by vishnu1991** [comment]
There is a series used by JD to show phone number. 
If we can extract the span>classname then we can get mobile numbers easily. 

Series is as below
**Number - span class="icon-XX"**
1 - icon-yz
2 -icon-wx
3 -icon-vu
4 -icon-ts
5 -icon-rq
6 -icon-po
7 -icon-nm
8 -icon-lk
9 -icon-ji
0 -icon-acb

**Comment by Alankar0416** [comment]
Yes, I had that in mind. But the issue is they can change the class name whenever they want and this will break then. Better to think of something concrete. The most foolproof solution is to use digit recognition on the image.

**Comment by vishnu1991** [comment]
yes i think the same. as the will surely change it.

**Comment by krishnamalireddy** [comment]
I'm not getting the phonenumbers. Can you tell me how to get phone numbers

**Comment by Alankar0416** [comment]
@krishnamalireddy  JD is now using svg's in place of actual numbers. That's why parsing is getting failed. There are couple of ways to get around this. 

Each svg's has a unique code which can be mapped - will fail if they change mapping again
Use a digit recognition over the svg.

Unfortunately I am not getting time to develop this. Will pick it up whenever I have some bandwidth.

**Comment by hrwndr** [comment]
@Alankar0416 Could you please demonstrate, how can we implement the numbers from svgs in code?

**Comment by AdityaMalireddy** [comment]
> @Alankar0416 Could you please demonstrate, how can we implement the numbers from svgs in code?

simple solution is instead of using .string use .find_all for phone number.

You will get random code of svg's convert them

**Comment by Alankar0416** [comment]
The issue is we can to keep a map of svg code and number but it JD can change it anytime.

**Comment by AdityaMalireddy** [comment]
Ha they can change it any time. If they have changed we have to decode it again. By the way they haven't changed it for a long time

**Comment by ketanshah79** [comment]
Thanks @Alankar0416 for sharing the code. 

Here is an array mapping I've used as a second pass on the csv file. I used the .find_all for phone number.

* '<bound method Tag.find_all of <a><b>' => '',
* '</b></a>>' => '',
* '<span class=""mobilesv icon-dc""></span>' => '',
* '<span class=""mobilesv icon-fe""></span>' => '',
* '<span class=""mobilesv icon-hg""></span>' => '',
* '<span class=""mobilesv icon-ba""></span>' => '-',
* '<span class=""mobilesv icon-acb""></span>' => '0',
* '<span class=""mobilesv icon-yz""></span>' => '1',
* '<span class=""mobilesv icon-wx""></span>' => '2',
* '<span class=""mobilesv icon-vu""></span>' => '3',
* '<span class=""mobilesv icon-ts""></span>' => '4',
* '<span class=""mobilesv icon-rq""></span>' => '5',
* '<span class=""mobilesv icon-po""></span>' => '6',
* '<span class=""mobilesv icon-nm""></span>' => '7',
* '<span class=""mobilesv icon-lk""></span>' => '8',
* '<span class=""mobilesv icon-ji""></span>' => '9',
* '<bound method Tag.find_all of <a>' => '',
* '</a>>' => '',

Attached is my php code.
[clean_csv.php.txt](https://github.com/Alankar0416/Justdial-Scrapper/files/2840899/clean_csv.php.txt)

**Comment by Alankar0416** [comment]
Great work @ketanshah79 
Haven't tried this code. Are you able to successfully map phone numbers with this additional script? If yes, I can add this into the original script to make things easy for everyone.

**Comment by ketanshah79** [comment]
Yes I did get it to work. Try running it on any CSV generated by your
python script

Thank you

**Comment by Dhiren-Biren** [comment]
only 10 data retrieving

**Comment by mps1305** [comment]
@Alankar0416 could you please post the code along with @ketanshah79 's changes?
Need to get justdial data for a college project.
Please guys, if either of you could do it, it'll be really helpful

Thanks!

**Comment by dvijparekh** [comment]
> @Alankar0416 could you please post the code along with @ketanshah79 's changes?
> Need to get justdial data for a college project.
> Please guys, if either of you could do it, it'll be really helpful
> 
> Thanks!

@mps1305 check  my forked repo i have made changes accordingly and its working just change url whichever you want

**Comment by mps1305** [comment]
hey @dvijparekh , it was working up until sometime back. then started getting this error. Any help in this regard would be highly appreciated!
"[WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond"

**Comment by dvijparekh** [comment]
it seems like justdial is blocking scraper to scrape working on it

**Comment by hrwndr** [comment]
I scrapped contact number successfully but the svg code changes everytime i run the script. How can i the correct svg code?
for example: i ran the script two times. Both the outputs were different.
1st time:
[https://ibb.co/Sx85zvb](url)
2nd time:
[https://ibb.co/34xBHrp](url)

**Comment by SuhailSaify** [comment]
Hey, I have written a script that will scrape phone numbers from any JustDail Business page.
It uses the info in CSS stylesheet to create a mapping between the strings assigned to each number.
The mapping is done every time you load a page, therefore it works for every business.

Please try this: 
https://github.com/SuhailSaify/Justdial-Scrapper

PS: it also scrapes other info along with Phone numbers.
(Working on July, 2019)

**Comment by krishnamalireddy** [comment]
I am getting urllib open timeout error. Is this code still working for anyone?

**Comment by abhi-ux** [comment]
can anyone update latest code here?

**Comment by SuhailSaify** [comment]
> can anyone update latest code here?

try this:
https://github.com/SuhailSaify/Justdial-Scrapper

**Comment by abhi-ux** [comment]
The phone number is not correct

**Comment by builditpossible-gs** [comment]
I am about to solve this issue, can anyone help me with this error - https://stackoverflow.com/questions/60875316/typeerror-string-indices-must-be-integers-when-getting-class-fro-span-tag-using

**Comment by dvijparekh** [comment]
> I am about to solve this issue, can anyone help me with this error - https://stackoverflow.com/questions/60875316/typeerror-string-indices-must-be-integers-when-getting-class-fro-span-tag-using

please share link url of just dial you are trying to scrape

**Comment by builditpossible-gs** [comment]
> > I am about to solve this issue, can anyone help me with this error - https://stackoverflow.com/questions/60875316/typeerror-string-indices-must-be-integers-when-getting-class-fro-span-tag-using
> 
> please share link url of just dial you are trying to scrape

Solved it brother. Thank you.

**Comment by builditpossible-gs** [comment]
There is another error though
`AttributeError: 'NoneType' object has no attribute 'text'` on line `return body.find('span', {'class':'mrehover'}).text.strip()` in `get_address`

**Comment by dvijparekh** [comment]
> There is another error though
> `AttributeError: 'NoneType' object has no attribute 'text'` on line `return body.find('span', {'class':'mrehover'}).text.strip()` in `get_address`

it means it is not able to find span tag having class mrehover so body.find is returning none which doesnt have any method or  attribute text() 
try below code and let me know what are you getting from it

`tesVar  = body.find('span', {'class':'mrehover'})`
print(`tesVar)

**Comment by alokm014** [comment]
Hey, use this method https://youtu.be/EkbF5JwuHqU
